README.md
# react-advanced-timefield

Advanced react time input field. Based out of the [Simple React time input field](https://github.com/antonfisher/react-simple-timefield).

[![Build Status](https://travis-ci.org/dotsub/react-advanced-timefield.svg?branch=master)](https://travis-ci.org/dotsub/react-advanced-timefield)
[![npm](https://img.shields.io/npm/dt/react-advanced-timefield.svg?colorB=brightgreen)](https://www.npmjs.com/package/react-advanced-timefield)
[![npm](https://img.shields.io/npm/v/react-advanced-timefield.svg?colorB=brightgreen)](https://www.npmjs.com/package/react-advanced-timefield)
[![GitHub license](https://img.shields.io/github/license/dotsub/react-advanced-timefield.svg)](https://github.com/dotsub/react-advanced-timefield/blob/master/LICENSE)

## Differences with react-simple-timefield

- Added milliseconds support with `showMillis` property
- Removed 24 hour limit (new limit is 99)
- Added up/down arrow support to increase or decrease based on cursor position
- Added className property to receive css classes
- Added TypeScript compatibility

## Installation
```bash
npm install --save react-advanced-timefield

#for React <16 use: npm install --save react-advanced-timefield@1
```

## Usage
```jsx
import TimeField from 'react-advanced-timefield';
...
<TimeField
    value={time}                       // {String}   required, format '00:00' or '00:00:00'
    onChange={(event, value) => {...}} // {Function} required
    input={<MyCustomInputElement />}   // {Element}  default: <input type="text" />
    colon=":"                          // {String}   default: ":"
    showSeconds                        // {Boolean}  default: false
    showMillis                         // {Boolean}  default: false
/>
```

## Real world example
```jsx
import TimeField from 'react-advanced-timefield';

class App extends React.Component {
  constructor(...args) {
    super(...args);

    this.state = {
      time: '12:34'
    };

    this.onTimeChange = this.onTimeChange.bind(this);
  }

  onTimeChange(event, time) {
    this.setState({time});
  }

  render() {
    const {time} = this.state;

    return (
      <TimeField value={time} onChange={this.onTimeChange} />
    );
  }
}
```

#### Run demo:
For running demo locally, replace:
```javascript
import TimeField from '../';
// to
import TimeField from '../src';
```
in `demo/index.tsx` file.

```bash
# run development mode
cd demo
npm run dev
```

#### Build:
```bash
npm test
npm run format
npm run build
```

## License
MIT License. Free use and change.


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file. See [standard-version](https://github.com/conventional-changelog/standard-version) for commit guidelines.

# [3.6.7](https://github.com/dotsub/react-advanced-timefield/compare/v3.5.5...v3.6.7) (2022-02-17)


### Features

* Dependency updates

# [3.5.5](https://github.com/dotsub/react-advanced-timefield/compare/v3.5.4...v3.5.5) (2020-03-05)


### Bug Fixes

* Event object getting lost on keyDown ([PR12](https://github.com/dotsub/react-advanced-timefield/pull/12))


# [3.5.4](https://github.com/dotsub/react-advanced-timefield/compare/v3.5.0...v3.5.4) (2020-01-20)


### Bug Fixes

* Typescript compatibility fix ([PR10](https://github.com/dotsub/react-advanced-timefield/pull/10))
* Change TS module from ESNext to CommonJs ([PR11](https://github.com/dotsub/react-advanced-timefield/pull/11))

# [3.5.0](https://github.com/dotsub/react-advanced-timefield/compare/v3.0.0...v3.5.0) (2020-01-20)


### Features

* Typescript compatibility ([PR9](https://github.com/dotsub/react-advanced-timefield/pull/9))

# [3.4.0](https://github.com/dotsub/react-advanced-timefield/compare/v3.0.0...v3.2.0) (2020-01-17)


### Features

* Add milliseconds support ([PR5](https://github.com/dotsub/react-advanced-timefield/pull/5))
* Remove 24 hour limit (new limit is 99) ([PR6](https://github.com/dotsub/react-advanced-timefield/pull/6))
* Add up/down arrow support ([PR7](https://github.com/dotsub/react-advanced-timefield/pull/7))
* Add className property ([PR8](https://github.com/dotsub/react-advanced-timefield/pull/8))


## [3.0.0](https://github.com/antonfisher/react-simple-timefield/compare/v2.0.3...v3.0.0) (2019-09-29)


### Bug Fixes

* [BREAKING CHANGES] add event to callback arguments (closes [#5](https://github.com/antonfisher/react-simple-timefield/issues/5)) ([5b5849c](https://github.com/antonfisher/react-simple-timefield/commit/5b5849c))
* remove react deprecation warning, replace componentWillReceiveProps call (closes [#11](https://github.com/antonfisher/react-simple-timefield/issues/11)) ([ad1f667](https://github.com/antonfisher/react-simple-timefield/commit/ad1f667))

## [2.0.3](https://github.com/antonfisher/react-simple-timefield/compare/v2.0.2...v2.0.3) (2019-04-10)


### Bug Fixes

* include build command to publish process ([600a320](https://github.com/antonfisher/react-simple-timefield/commit/600a320))



## [2.0.2](https://github.com/antonfisher/react-simple-timefield/compare/v2.0.1...v2.0.2) (2019-04-10)


### Bug Fixes

* user can type letter when component first loads (closes [#9](https://github.com/antonfisher/react-simple-timefield/issues/9)) ([8f51c70](https://github.com/antonfisher/react-simple-timefield/commit/8f51c70))



<a name="2.0.1"></a>
## [2.0.1](https://github.com/antonfisher/react-simple-timefield/compare/v2.0.0...v2.0.1) (2018-11-23)


### Bug Fixes

* check if undefined for 'value' and 'onChange' (issue [#8](https://github.com/antonfisher/react-simple-timefield/issues/8)) ([08e6c2e](https://github.com/antonfisher/react-simple-timefield/commit/08e6c2e))
* update deps ([e7f7d20](https://github.com/antonfisher/react-simple-timefield/commit/e7f7d20))



<a name="2.0.0"></a>
# [2.0.0](https://github.com/antonfisher/react-simple-timefield/compare/v1.3.2...v2.0.0) (2018-02-01)


### Bug Fixes

* remove double usage of same code ([ee99ce4](https://github.com/antonfisher/react-simple-timefield/commit/ee99ce4))


### Features

* React v16 support ([127217b](https://github.com/antonfisher/react-simple-timefield/commit/127217b))


tests/setup.ts
import {configure} from 'enzyme';
import Adapter from '@wojtekmaj/enzyme-adapter-react-17';

configure({
  adapter: new Adapter()
});


tests/unit.test.ts
import {isNumber, formatTimeItem, validateTimeAndCursor, increment} from '../src/index';

describe('#isNumber()', () => {
  test('should return true for numbers', () => {
    expect(isNumber(0)).toBe(true);
    expect(isNumber(1)).toBe(true);
    expect(isNumber('0')).toBe(true);
    expect(isNumber('1')).toBe(true);
  });

  test('should return false for not numbers', () => {
    expect(isNumber('')).toBe(false);
    expect(isNumber('a')).toBe(false);
    expect(isNumber(' 1')).toBe(false);
  });
});

describe('#formatTimeItem()', () => {
  test('should return default value if string is empty', () => {
    expect(formatTimeItem()).toBe('00');
  });

  test('should return formated value', () => {
    expect(formatTimeItem('1')).toBe('10');
    expect(formatTimeItem('11')).toBe('11');
    expect(formatTimeItem('111')).toBe('11');
    expect(formatTimeItem(2)).toBe('20');
  });
});

describe('#validateTimeAndCursor()', () => {
  const DF = '00:00:00.000';

  test('should return an array', () => {
    const res = validateTimeAndCursor(true, true, '', DF, ':', 0);
    expect(res).toBeInstanceOf(Array);
    expect(res).toHaveLength(2);
    expect(res).toEqual([DF, 0]);
  });

  test('should handle "showSeconds" option', () => {
    expect(validateTimeAndCursor(true, false, '12:34:56', DF)[0]).toEqual('12:34:56');
    expect(validateTimeAndCursor(true, false, '12:34', DF)[0]).toEqual('12:34:00');
    expect(validateTimeAndCursor(false, false, '12:34:56', DF)[0]).toEqual('12:34');
    expect(validateTimeAndCursor(false, false, '12:34', DF)[0]).toEqual('12:34');
  });

  test('should handle "showMillis" option', () => {
    expect(validateTimeAndCursor(true, true, '12:34:56.789', DF)[0]).toEqual('12:34:56.789');
    expect(validateTimeAndCursor(true, true, '12:34', DF)[0]).toEqual('12:34:00.000');
    expect(validateTimeAndCursor(false, true, '12:34:56.789', DF)[0]).toEqual('12:34');
    expect(validateTimeAndCursor(false, true, '12:34', DF)[0]).toEqual('12:34');
  });

  test('should handle "colon" option', () => {
    expect(validateTimeAndCursor(true, true, '12-34-56.789', DF, '-')[0]).toEqual('12-34-56.789');
    expect(validateTimeAndCursor(true, true, '12-34', DF, '-')[0]).toEqual('12-34-00.000');
    expect(validateTimeAndCursor(true, false, '12-34-56', DF, '-')[0]).toEqual('12-34-56');
    expect(validateTimeAndCursor(true, false, '12-34', DF, '-')[0]).toEqual('12-34-00');
    expect(validateTimeAndCursor(false, false, '12-34-56', DF, '-')[0]).toEqual('12-34');
    expect(validateTimeAndCursor(false, false, '12-34', DF, '-')[0]).toEqual('12-34');
  });

  test('should validate hours', () => {
    expect(validateTimeAndCursor(false, false, '00:00', DF)[0]).toEqual('00:00');
    expect(validateTimeAndCursor(false, false, '12:00', DF)[0]).toEqual('12:00');
    expect(validateTimeAndCursor(false, false, '23:00', DF)[0]).toEqual('23:00');
    expect(validateTimeAndCursor(false, false, '24:00', DF)[0]).toEqual('24:00');
    expect(validateTimeAndCursor(false, false, '99:00', DF)[0]).toEqual('99:00');
    expect(validateTimeAndCursor(false, false, '1:00', DF)[0]).toEqual('10:00');
    expect(validateTimeAndCursor(false, false, '24:00', '21:00')[0]).toEqual('24:00');
  });

  test('should validate minutes', () => {
    expect(validateTimeAndCursor(false, false, '12:00', DF)[0]).toEqual('12:00');
    expect(validateTimeAndCursor(false, false, '12:30', DF)[0]).toEqual('12:30');
    expect(validateTimeAndCursor(false, false, '12:59', DF)[0]).toEqual('12:59');
    expect(validateTimeAndCursor(false, false, '12:60', DF)[0]).toEqual('12:00');
    expect(validateTimeAndCursor(false, false, '12:1', DF)[0]).toEqual('12:10');
  });

  test('should validate seconds', () => {
    expect(validateTimeAndCursor(true, false, '12:00:00', DF)[0]).toEqual('12:00:00');
    expect(validateTimeAndCursor(true, false, '12:00:30', DF)[0]).toEqual('12:00:30');
    expect(validateTimeAndCursor(true, false, '12:00:59', DF)[0]).toEqual('12:00:59');
    expect(validateTimeAndCursor(true, false, '12:00:60', DF)[0]).toEqual('12:00:00');
    expect(validateTimeAndCursor(true, false, '12:00:1', DF)[0]).toEqual('12:00:10');
  });

  test('should validate millis', () => {
    expect(validateTimeAndCursor(true, true, '12:00:00.000', DF)[0]).toEqual('12:00:00.000');
    expect(validateTimeAndCursor(true, true, '12:00:00.999', DF)[0]).toEqual('12:00:00.999');
    expect(validateTimeAndCursor(true, true, '12:00:00.09', DF)[0]).toEqual('12:00:00.090');
    expect(validateTimeAndCursor(true, true, '12:00:00.9', DF)[0]).toEqual('12:00:00.900');
    expect(validateTimeAndCursor(true, true, '12:00:00.', DF)[0]).toEqual('12:00:00.000');
  });
});

describe('#increment()', () => {
  test('should return original value incremented by 1', () => {
    expect(increment('00', 1)).toBe('01');
    expect(increment('000', 1)).toBe('001');
  });

  test('should return minimum value of 0', () => {
    expect(increment('00', -1)).toBe('00');
    expect(increment('23', -34)).toBe('00');
    expect(increment('000', -1)).toBe('000');
    expect(increment('023', -34)).toBe('000');
  });

  test('should return original value incremented by 15', () => {
    expect(increment('06', 15)).toBe('21');
    expect(increment('006', 15)).toBe('021');
  });
});


demo/webpack.config.dev.js
const path = require('path');
const HtmlWebpackPlugin = require('html-webpack-plugin');
const {CleanWebpackPlugin} = require('clean-webpack-plugin');

const outputPath = path.join(__dirname, 'build-dev');

module.exports = {
  mode: 'development',
  context: __dirname,
  entry: './index.js',
  output: {
    publicPath: '',
    path: outputPath,
    filename: 'bundle.js'
  },
  module: {
    rules: [
      {
        test: /\.js$/,
        use: 'babel-loader',
        exclude: /node_modules/
      },
      {
        test: /\.tsx?$/,
        use: 'ts-loader',
        exclude: /node_modules/
      }
    ]
  },
  resolve: {
    extensions: ['.tsx', '.js']
  },
  plugins: [
    new CleanWebpackPlugin(),
    new HtmlWebpackPlugin({
      hash: true,
      inject: true,
      template: 'index.html'
    })
  ],
  devtool: 'inline-source-map',
  devServer: {
    static: outputPath
  }
};


demo/index.js
import 'babel-polyfill';
import React from 'react';
import ReactDOM from 'react-dom';
import TextField from '@material-ui/core/TextField';

// import TimeField from '../';
import TimeField from '../src';

class App extends React.Component {
  constructor() {
    super();

    this.state = {
      time: '12:34',
      timeSeconds: '12:34:56',
      timeMillis: '12:34:56.789',
      timeSecondsCustomColon: '12-34-56'
    };

    this.onTimeChange = this.onTimeChange.bind(this);
  }

  onTimeChange(event, value) {
    const newTime = value.replace(/-/g, ':');
    const time = newTime.substr(0, 5);
    const timeSeconds = newTime.padEnd(8, this.state.timeSeconds.substr(5, 3));
    const timeMillis = timeSeconds.padEnd(12, this.state.timeMillis.substr(8, 4));
    const timeSecondsCustomColon = timeSeconds.replace(/:/g, '-');

    this.setState({time, timeSeconds, timeMillis, timeSecondsCustomColon});
  }

  render() {
    const {time, timeSeconds, timeMillis, timeSecondsCustomColon} = this.state;

    return (
      <section className="container">
        <h2>Default input:</h2>
        <section>
          <TimeField
            value={time}
            onChange={this.onTimeChange}
            onKeyDown={this.onTimeChange}
            style={{
              border: '2px solid #666',
              fontSize: 42,
              width: 107,
              padding: '5px 8px',
              color: '#333',
              borderRadius: 3
            }}
          />
        </section>
        <h2>Show seconds:</h2>
        <section>
          <TimeField
            showSeconds
            value={timeSeconds}
            onChange={this.onTimeChange}
            style={{
              border: '2px solid #666',
              fontSize: 42,
              width: 167,
              padding: '5px 8px',
              color: '#333',
              borderRadius: 3
            }}
          />
        </section>
        <h2>Show millis:</h2>
        <section>
          <TimeField
            showSeconds
            showMillis
            value={timeMillis}
            onChange={this.onTimeChange}
            style={{
              border: '2px solid #666',
              fontSize: 42,
              width: 255,
              padding: '5px 8px',
              color: '#333',
              borderRadius: 3
            }}
          />
        </section>
        <h2>Custom colon:</h2>
        <section>
          <TimeField
            showSeconds
            colon="-"
            value={timeSecondsCustomColon}
            onChange={this.onTimeChange}
            style={{
              border: '2px solid #666',
              fontSize: 42,
              width: 180,
              padding: '5px 8px',
              color: '#333',
              borderRadius: 3
            }}
          />
        </section>
        <h2>React Material-UI:</h2>
        <section>
          <div style={{marginRight: 20}}>
            <TimeField
              showSeconds
              value={timeSeconds}
              onChange={this.onTimeChange}
              style={{width: 88}}
              input={<TextField label="Name" value={timeSeconds} variant="outlined" />}
            />
          </div>
        </section>
        <h2>With CSS class:</h2>
        <section>
          <div style={{marginRight: 20}}>
            <TimeField
              value={time}
              onChange={this.onTimeChange}
              className="custom-class"
              style={{textAlign: 'right'}}
            />
          </div>
        </section>
      </section>
    );
  }
}

ReactDOM.render(<App />, document.getElementById('app'));
